src/libYARP_OS/include/yarp/os/ThreadStats.h

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright: (C) 2017 iCub Facility
+ * Author: Diego Ferigo
+ * CopyPolicy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+ */
+
+#ifndef YARP_THREAD_STATS_H
+#define YARP_THREAD_STATS_H
+
+#include <cassert>
+#include <cmath>
+#include <map>
+#include <yarp/os/ConstString.h>
+#include <yarp/os/LogStream.h>
+#include <yarp/os/Mutex.h>
+#include <yarp/os/Time.h>
+
+/*
+    Consider:
+    * |---------| Thread's or DUT's period
+    * x           Time unit taken by the real computation
+    * -           Time unit not taken by the real computation
+
+    tick  tock
+    v     v
+    |xxxxxx---|xxxx-----|xxxxx----|xxxxxx---|
+     <--------><--------><--------><-------->   dT
+     <---->    <-->      <--->     <---->       elapsed
+*/
+
+namespace yarp {
+namespace os {
+class ThreadStats;
+class MultipleThreadStats;
+const unsigned WINDOW_LENGTH_DEFAULT = 60;
+const unsigned BINS_PITCH_MS_DEFAULT = 5;
+const unsigned MAX_JITTER_MS_DEFAULT = 100;
+} // namespace os
+} // namespace yarp
+
+class YARP_OS_API yarp::os::ThreadStats {
+private:
+    // All times are in seconds and must be positive and non-zero
+
+    // Constant values
+    // ===============
+    double window_length;       // Lenght of the window for calculating statistic features
+    double period;              // Ideal period
+    unsigned cycles_per_window; // Number of cycles that fit in a selected window
+    unsigned maxJitter_ms;      // Maximum jitter allowed
+
+    // Instantaneous values
+    // ====================
+    double dT;        // Real period
+    double tick_time; // Time acquired during the tick phase
+    double tock_time; // Time acquired during the tock phase
+    double elapsed;   // Effective computational time taken by the thread
+
+    // Variables that keep data among cycles into the window_length
+    // ============================================================
+    unsigned cycle_count; // Counter of elapsed cycles
+    double time_total;    // Time passed from the last reset (sum of dT until window_length)
+    double elapsed_total; // Effective time passed from the last reset (sum of elapsed)
+    double elapsed_mu;    // Mean of the elapsed time over cycles inside the window_length
+    double elapsed_M2;    // Variable used for the online calculation of the STD of the elapsed time
+    double dT_mu;         // Mean of the real period
+    double dT_M2;         // Variable used for the online calculation of the STD of the real period
+
+    // Advanced statistics: histogram
+    // ==============================
+    unsigned bins_pitch_ms;
+    std::map<unsigned, unsigned> jitter_pdf;
```

Comment:
Make sure this list is allocated before the thread starts running and deallocated after the thread stopped. If you modify the size of the list dynamically during the thread execution there may be a performance hit.